### PR TITLE
Support attributes for items of Dataset

### LINKED ISSUE #666: Attributes for items in Dataset
The addition of first attributes in real examples highlighted an issue with the current concept: So far, attributes are global for a `Dataset`. When inserting a `DataArray` into a dataset the implementation compares attributes with the same name and fails on mismatch. This is not a usable approach. Therefore, we must support attributes for each array (data item) in a dataset.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #666.

This implementation does *not* map dataset attributes into the attribute dict of its items. This is unlike coords, labels, or masks. There are two reasons:
- Complications and ambiguities when implementing operations on dataset, which are based on applying the operation to all items --- if the operation preserves attributes then the dataset attribute that was mapped into the input items attr dicts would get copied as new attribute for each item of the output.
- Danger of unwillingly changing dataset attributes when modifying item attributes, since with the existing syntax there is no way of distinguishing item attributes from dataset attributes:
  ```python
  d['a'].attrs['facility'] = 'xxx' # is this changing an attribute of `d`?
  ```

### CODE REVIEW

File: core/test/attributes_test.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  d1["a"].attrs().set("a_attr", scalar);
+  d1.attrs().set("dataset_attr", scalar);
+
+  Dataset d2;
+  d2.setData("a", varX);
+  d2["a"].attrs().set("a_attr", varX);
+  d2.attrs().set("dataset_attr", varX);
+
+  auto result(d1);
+  result += d2;
```
Comment (nvaytet) [inline core/test/attributes_test.cpp]: Maybe worth checking here some intermediate results? The labels are just compared right? But what are the attributes of `resut` at this point here? I guess it is not replaced if it already exists? So you could also check here that
```C++
  EXPECT_EQ(result.attrs()["dataset_attr"], scalar);
  EXPECT_EQ(result["a"].attrs()["a_attr"], scalar);
```
In addition, what happens if you have
```C++
  Dataset d1;
  d1.setData("a", varX);
  d1["a"].attrs().set("a_attr1", scalar);
  d1.attrs().set("dataset_attr1", scalar);

  Dataset d2;
  d2.setData("a", varX);
  d2["a"].attrs().set("a_attr2", varX);
  d2.attrs().set("dataset_attr2", varX);

  auto result(d1);
  result += d2;
```
at this point, does `result` hold both `"dataset_attr1"` and `"dataset_attr2"`?

File: core/test/attributes_test.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  result -= d2;
+  result *= d2;
+  result /= d2;
+  ASSERT_TRUE(result.attrs().contains("dataset_attr"));
+  EXPECT_EQ(result.attrs()["dataset_attr"], scalar);
+  ASSERT_TRUE(result["a"].attrs().contains("a_attr"));
+  EXPECT_EQ(result["a"].attrs()["a_attr"], scalar);
+}
+
```
Comment (nvaytet) [inline core/test/attributes_test.cpp]: Maybe you could also have a test where the operations are still being applied even if the attribute keys don't match?
```C++
  Dataset d1;
  d1.setData("a", varX);
  d1["a"].attrs().set("a_attr1", scalar);
  d1.attrs().set("dataset_attr1", scalar);

  Dataset d2;
  d2.setData("a", varX);
  d2["a"].attrs().set("a_attr2", varX);
  d2.attrs().set("dataset_attr2", varX);

  auto result = d1 + d2;
```
or something like that, but then maybe this is already tested elsewhere? I tried having a quick look around `core` but I did not really find anything.

File: core/test/attributes_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply core/test/attributes_test.cpp]: Not sure what is gained by checking intermediate results, unless you expect one operator to magically restore things that a previous one removed?

Attributes from `d2` are always ignored, regardless of the name. I will clarfiy this in the test.

File: core/test/attributes_test.cpp (hunk 2 above)
Comment (SimonHeybrock) [reply core/test/attributes_test.cpp]: See above, this is just ignoring attributes from the second one.

Comment (SimonHeybrock) [reply core/test/attributes_test.cpp]: Also, non-in-place binary operations just drop all attributes.

Comment (nvaytet) [reply core/test/attributes_test.cpp]: > Also, non-in-place binary operations just drop all attributes.

Is there a test that illustrates this? e.g.
```C++
auto result = d1 + d2;
ASSERT_FALSE(result["a"].attrs().contains("a_attr1"));
ASSERT_FALSE(result["a"].attrs().contains("a_attr2"));
```

File: core/test/attributes_test.cpp (hunk 1 above)
Comment (nvaytet) [reply core/test/attributes_test.cpp]: Ok, making the test a bit clearer would work

**Pull request merged**